src/net_processing.cpp

```diff
@@ -760,6 +764,12 @@ void PeerLogicValidation::NewPoWValidBlock(const CBlockIndex *pindex, const std:
     bool fWitnessEnabled = IsWitnessEnabled(pindex->pprev, Params().GetConsensus());
     uint256 hashBlock(pblock->GetHash());
 
+    {
```

Comment:
These are the types of changes that make me nervous about shared_ptrs, because it gets easy to start stashing things that have no obvious lifetime constraint.

I was suggesting that you hold the shared_ptr in the PeerLogicValidation class. You then use a global weak_ptr which is overwritten by each new block. Then at any random time you can use std::shared_ptr\<const CBlock\> bestblock = shared_ptrglobal_last_block.lock().

That way you can extend the lifetime of the whatever block is current, but it's also obvious when it will be destroyed (with the PeerLogicValidation destruction).